File: ts/src/Mercator.ts
import Constant from './Constant';
import LngLat from './LngLat';

export default class Mercator {
  private x: number;
  private y: number;

  private lng;
  private lat;

  public static formLngLat(lngLat: LngLat): Mercator {
    return new Mercator(lngLat.getX(), lngLat.getY());
  }

  public static toLngLat(xy: Mercator, clip = false): LngLat {
    return new LngLat(xy.getLng(), xy.getLat(), clip);
  }

  constructor(x: number, y: number) {
    this.setX(x);
    this.setY(y);
  }

  public getX() {
    return this.x;
  }

  public setX(x: number) {
    this.x = x;

    this.lng = (this.x * Constant.R2D) / Constant.RE;
    return this;
  }

  public getY() {
    return this.y;
  }

  public getLng() {
    return this.lng;
  }

  public getLat() {
    return this.lat;
  }

  public setY(y: number) {
    this.y = y;

    this.lat = (Math.PI * 0.5 - 2.0 * Math.atan(Math.exp(-y / Constant.RE))) * Constant.R2D;
    return this;
  }

  public equals(m: Mercator) {
    if (m === this) {
      return true;
    }
    return Constant.closeTo(this.getX(), m.getX()) && Constant.closeTo(this.getY(), m.getY());
  }
}

File: java/src/main/java/com/sakitam/mercantile/Mercator.java
package com.sakitam.mercantile;

import java.io.Serializable;

public class Mercator implements Serializable {
    private static final long serialVersionUID = 1L;

    private double x;
	private double y;

	private double lng;

	private double lat;

	public Mercator() {
		super();
	}

	public Mercator(double x, double y) {
		super();
		this.setX(x);
		this.setY(y);
	}

	public double getX() {
		return x;
	}

	public void setX(double x) {
		this.x = x;

		this.lng = this.x * Constant.R2D / Constant.RE;
	}

	public double getY() {
		return y;
	}

	public double getLng() {
		return lng;
	}

	public double getLat() {
		return lat;
	}

	public void setY(double y) {
		this.y = y;

		this.lat = ((Math.PI * 0.5) - 2.0 * Math.atan(Math.exp(-y / Constant.RE))) * Constant.R2D;
	}

	public static Mercator formLngLat(LngLat lngLat) {
		return new Mercator(lngLat.getX(), lngLat.getY());
	}

	public static LngLat toLngLat(Mercator xy, boolean clip) {
		return new LngLat(xy.getLng(), xy.getLat(), clip);
	}
}
